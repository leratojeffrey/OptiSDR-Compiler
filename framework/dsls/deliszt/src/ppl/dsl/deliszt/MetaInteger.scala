package ppl.dsl.deliszt

/**
 * author: Michael Wu (mikemwu@stanford.edu)
 * last modified: 04/28/2011
 *
 * Pervasive Parallelism Laboratory (PPL)
 * Stanford University
 */

/*
 * MetaInteger.scala
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

abstract class IntM {
  type Suc <: IntM
  type Add[N <: IntM] <: IntM
}

object EnsureSize {
  implicit def EnsureSizeZero[T <: IntM] =
  new EnsureSize[Zero,Succ[T]] {
    def get(n : Int) : Int = n
  }

  implicit def EnsureSizeSucc[T <: IntM, TT <: IntM](implicit fn : EnsureSize[TT,T]) =
  new EnsureSize[Succ[TT],Succ[T]] {
    def get(n : Int) : Int = fn.get(n+1)
  }
}

//meta-function(?)(what is the term for this thing?)
//if IDX is less than SIZE, then an implicit object of type EnsureSize[IDX,SIZE] will
//exist with a idx function that returns the number IDX as an integer
abstract class EnsureSize[TT <: IntM,T <: IntM] {
  def get(n : Int) : Int
  def idx = get(0)
}

object MVal {
  implicit val MValZero = new MVal[Zero] {
    val value = 0
  }

  implicit def MValSucc[T <: IntM](implicit f: MVal[T]) = new MVal[Succ[T]] {
    val value = f.value + 1
  }
}

abstract class MVal[T] {
  val value: Int
}

class Zero extends IntM {
  type Suc = Succ[Zero]
  type Add[N <: IntM] = N
}

class Succ[N <: IntM] extends IntM {
  type Suc = Succ[Succ[N]]
  type Add[M <: IntM] = (N#Add[M])#Suc
}

object MetaInteger {
  type +[N1 <: IntM, N2 <: IntM] = N1#Add[N2]

  //Scala doesn't provide meta-integers like C++'s typename<int N>
  //so we have to define some integers here
  //this is a simplification of:
  //http://michid.wordpress.com/code/meta-programming-with-scala-part-i-addition/
  //somelike like Church encoding but the numbers are themselves objects and not
  //functions for simplicity since we don't need to define addition on them
  //as far I can tell ... if necessary I suppose we can define multiplcation
  //on these if we needed numbers higher than 9
  type _0 = Zero
  type _1 = Succ[_0]
  type _2 = Succ[_1]
  type _3 = Succ[_2]
  type _4 = Succ[_3]
  type _5 = Succ[_4]
  type _6 = Succ[_5]
  type _7 = Succ[_6]
  type _8 = Succ[_7]
  type _9 = Succ[_8]
  type _10 = Succ[_9]
  type _11 = Succ[_10]
  type _12 = Succ[_11]
  type _13 = Succ[_12]
  type _14 = Succ[_13]
  type _15 = Succ[_14]
  type _16 = Succ[_15]
  type _17 = Succ[_16]
  type _18 = Succ[_17]
  type _19 = Succ[_18]
  type _20 = Succ[_19]
  type _21 = Succ[_20]
  type _22 = Succ[_21]
  
  //for convience we provide objects representing each number as well
  //so you can avoid passing parameter arguments
  val _0 : _0 = new _0
  val _1 : _1 = new _1
  val _2 : _2 = new _2
  val _3 : _3 = new _3
  val _4 : _4 = new _4
  val _5 : _5 = new _5
  val _6 : _6 = new _6
  val _7 : _7 = new _7
  val _8 : _8 = new _8
  val _9 : _9 = new _9
  val _10 : _10 = new _10
  val _11 : _11 = new _11
  val _12 : _12 = new _12
  val _13 : _13 = new _13
  val _14 : _14 = new _14
  val _15 : _15 = new _15
  val _16 : _16 = new _16
  val _17 : _17 = new _17
  val _18 : _18 = new _18
  val _19 : _19 = new _19
  val _20 : _20 = new _20
  val _21 : _21 = new _21
  val _22 : _22 = new _22

  def MIntDepth[T <: IntM]()(implicit n: MVal[T]) = n.value
  def MIntDepth[T <: IntM](m : T)(implicit n: MVal[T]) = n.value
}
