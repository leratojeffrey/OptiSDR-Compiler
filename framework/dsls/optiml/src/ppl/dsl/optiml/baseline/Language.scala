package ppl.dsl.optiml.baseline

object LanguageDefs {
  val INITIAL_SEED = 100
  var randRef = new scala.util.Random(INITIAL_SEED)
  var _time = 0L
  
  def random(): Double = randRef.nextDouble
    
  def abs(d: Double) = implicitly[Numeric[Double]].abs(d) //java.lang.Math.abs(d)
  
  def sum[B:Numeric:Manifest](st: Int, en: Int)(f: Int => B) = {
    var acc = f(st)
    for (i <- st+1 until en) {
      acc = implicitly[Numeric[B]].plus(acc,f(i))
    }
    acc
  }
  
  def tic() {
    _time = System.currentTimeMillis()
  }
  
  def toc() {
    val now = System.currentTimeMillis()
    println("elapsed: " + (now-_time)/1000.0)
  }
}