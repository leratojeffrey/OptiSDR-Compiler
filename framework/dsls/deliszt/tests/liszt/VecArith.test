import Liszt.Language._
import Liszt.MetaInteger._


@lisztcode
object VecArith {
	def main() {
		val a = Vec(1,2,3)
		val b = Vec(4.,5.,6.)
		val c = Vec(7.,8.,9.)
		val d = b + c
		val e = Vec(7,8,9) + a
		
		val f = 2.0 * b
		val g = b * 2.0
		val h = b * b
		
		val i = 2.0 / b
		val j = b / 2.0
		val k = b / b
		
		val l = Vec(Vec(1,2),Vec(3,4))
		val m = Vec(false,true)
		val n = b + b
		
		val o = Vec(true)
		val p = Vec("astring")
		val q = b - b
		
		val r = Vec(b.x,b.y,b.z)
		val s = b(_0)
		val t = Vec(1,2,3,4).w
		val u = dot(b,c)
		val v = cross(b,c)
		
		Print(a," ",b," ",c," ",d," ",e," ",f," ",g," ",h," ",i," ",j," ",k," ",l," ",m," ",n," ",o," ",p," ",q," ",r," ",s," ",t," ",u," ",v)

		val v1a = Vec(2.)
		val v1b = Vec(3.)
		val v2a = Vec(2.,3.)
		val v2b = Vec(5.,7.)
		val v3a = Vec(2.,3.,5.)
		val v3b = Vec(7.,11.,13.)
		val v4a = Vec(2.,3.,5.,7.)
		val v4b = Vec(11.,13.,17.,19.)
		val dot1 = dot(v1a, v1b)
		val dot2 = dot(v2a, v2b)
		val dot3 = dot(v3a, v3b)
		val dot4 = dot(v4a, v4b)
		Print(dot1, " ", dot2, " ", dot3, " ", dot4)  // 6, 10 + 21 = 31, 14 + 33 + 65 = 112, 22 + 39 + 85 + 133 = 279
		val normal1 = normalize(Vec(1.)) // (1)
		val normal2 = normalize(Vec(3.,4.))  // sqrt(9, 16) = 5, so (.6,.8)
		val normal3 = normalize(Vec(1.,4.,8.))  // sqrt(1 + 16 + 64) = 9, so (.111, .444, .888)
		val normal4 = normalize(Vec(1.,3.,3.,9.)) // sqrt(1 + 9 + 9 + 81) = 10, so (.1, .3, .3, .9)
		Print(normal1, " ", normal2, " ", normal3, " ", normal4)

		// Spec claims that min and max exist for Vectors
		val min1 = Vec(3) min Vec(1)  // (1)
		val min2 = Vec(1, 10) min Vec(5, 2)  // (1, 2)
		val min3 = Vec(1, 10, 15) min Vec(0, 20, 10) // (0, 10, 10)
		val min4 = Vec(0, 10, 5, 40) min Vec(5, 5, 5, 45) // (0, 5, 5, 40))
		Print(min1, " ", min2, " ", min3, " ", min4)

		val max1 = Vec(3) max Vec(1)  // (1)
		val max2 = Vec(1, 10) max Vec(5, 2)  // (1, 2)
		val max3 = Vec(1, 10, 15) max Vec(0, 20, 10) // (0, 10, 10)
		val max4 = Vec(0, 10, 5, 40) max Vec(5, 5, 5, 45) // (0, 5, 5, 40))
		
		val len = length(Vec(3.f,4.f))
		Print(max1, " ", max2, " ", max3, " ", max4, " ", len)
	}
}
