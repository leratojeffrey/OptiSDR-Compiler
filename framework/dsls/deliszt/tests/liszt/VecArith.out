0 liszt: (1,2,3) (4,5,6) (7,8,9) (11,13,15) (8,10,12) (8,10,12) (8,10,12) (16,25,36) (0.5,0.4,0.333333) (2,2.5,3) (1,1,1) ((1,2),(3,4)) (0,1) (8,10,12) (1) (astring) (0,0,0) (4,5,6) 4 4 122 (-3,6,-3)
1 liszt: 6 31 112 279
2 liszt: (1) (0.6,0.8) (0.111111,0.444444,0.888889) (0.1,0.3,0.3,0.9)
3 liszt: (1) (1,2) (0,10,10) (0,5,5,40)
4 liszt: (3) (5,10) (1,20,15) (5,10,5,45) 5
